File: android/src/main/kotlin/com/example/experiment_flutter/codec/UserCodec.kt
package com.example.experiment_flutter.codec

import android.util.Log
import com.amplitude.experiment.ExperimentUser
import com.example.experiment_flutter.toJSONObject
import com.example.experiment_flutter.toMap
import org.json.JSONObject

/**
 * Codec for serializing and deserializing ExperimentUser objects to/from Maps
 * for method channel communication.
 */
object UserCodec {
    /**
     * Parses an ExperimentUser from a Map representation.
     * @param userMap The map containing user data
     * @return An ExperimentUser object parsed from the map, or null if userMap is null
     */
    fun fromMap(userMap: Map<String, Any>?): ExperimentUser? {
        return userMap?.let { userArgs ->
            val builder = ExperimentUser.Builder()
            (userArgs["deviceId"] as? String)?.let { builder.deviceId(it) }
            (userArgs["userId"] as? String)?.let { builder.userId(it) }
            (userArgs["country"] as? String)?.let { builder.country(it) }
            (userArgs["city"] as? String)?.let { builder.city(it) }
            (userArgs["region"] as? String)?.let { builder.region(it) }
            (userArgs["dma"] as? String)?.let { builder.dma(it) }
            (userArgs["language"] as? String)?.let { builder.language(it) }
            (userArgs["platform"] as? String)?.let { builder.platform(it) }
            (userArgs["version"] as? String)?.let { builder.version(it) }
            (userArgs["os"] as? String)?.let { builder.os(it) }
            (userArgs["deviceModel"] as? String)?.let { builder.deviceModel(it) }
            (userArgs["deviceBrand"] as? String)?.let { builder.deviceBrand(it) }
            (userArgs["deviceManufacturer"] as? String)?.let { builder.deviceManufacturer(it) }
            (userArgs["carrier"] as? String)?.let { builder.carrier(it) }
            (userArgs["library"] as? String)?.let { builder.library(it) }

            // Decode userProperties from JSON string
            (userArgs["userProperties"] as? String)?.let { jsonString ->
                val jsonObject = JSONObject(jsonString)
                val userPropsMap = jsonObject.toMap() as Map<String, Any>
                builder.userProperties(userPropsMap)
            }

            // Decode groups from JSON string (Map<String, List<String>>)
            (userArgs["groups"] as? String)?.let { jsonString ->
                val jsonObject = JSONObject(jsonString)
                val groupsMap = jsonObject.toMap().mapValues { (_, value) ->
                    (value as? List<*>)?.map { it.toString() }?.toSet() ?: emptySet()
                } as Map<String, Set<String>>
                builder.groups(groupsMap)
            }

            // Decode groupProperties from JSON string (Map<String, Map<String, Map<String, Any?>>>)
            (userArgs["groupProperties"] as? String)?.let { jsonString ->
                val jsonObject = JSONObject(jsonString)
                val groupPropsMap = jsonObject.toMap().mapValues { (_, value) ->
                    (value as? Map<*, *>)?.mapValues { (_, innerValue) ->
                        (innerValue as? Map<*, *>)?.mapValues { it.value } as? Map<String, Any?>
                    } as? Map<String, Map<String, Any?>>
                } as Map<String, Map<String, Map<String, Any?>>>
                builder.groupProperties(groupPropsMap)
            }

            builder.build()
        }
    }

    /**
     * Converts a native Android ExperimentUser object to a Map that can be serialized
     * and sent to Flutter via method channels.
     * @param user The user to convert, or null
     * @return A map representation of the user, or empty map if user is null
     */
    fun toMap(user: ExperimentUser?): Map<String, Any?> {
        if (user == null) {
            return emptyMap()
        }

        val map = mutableMapOf<String, Any?>()
        user.deviceId?.let { map["deviceId"] = it }
        user.userId?.let { map["userId"] = it }
        user.country?.let { map["country"] = it }
        user.city?.let { map["city"] = it }
        user.region?.let { map["region"] = it }
        user.dma?.let { map["dma"] = it }
        user.language?.let { map["language"] = it }
        user.platform?.let { map["platform"] = it }
        user.version?.let { map["version"] = it }
        user.os?.let { map["os"] = it }
        user.deviceModel?.let { map["deviceModel"] = it }
        user.deviceBrand?.let { map["deviceBrand"] = it }
        user.deviceManufacturer?.let { map["deviceManufacturer"] = it }
        user.carrier?.let { map["carrier"] = it }
        user.library?.let { map["library"] = it }

        // Encode userProperties to JSON string
        user.userProperties?.let { userProps ->
            try {
                val jsonObject = userProps.toJSONObject()
                map["userProperties"] = jsonObject?.toString()
            } catch (e: Exception) {
                Log.e("ExperimentFlutter", "Error encoding userProperties: ${e.message}")
            }
        }

        // Encode groups to JSON string (Map<String, List<String>>)
        user.groups?.let { groups ->
            try {
                // Convert Set<String> to List<String> for Dart compatibility
                val groupsAsList = groups.mapValues { (_, value) ->
                    value.toList()
                }
                val jsonObject = groupsAsList.toJSONObject()
                map["groups"] = jsonObject?.toString()
            } catch (e: Exception) {
                Log.e("ExperimentFlutter", "Error encoding groups: ${e.message}")
            }
        }

        // Encode groupProperties to JSON string (Map<String, Map<String, Map<String, Any?>>>)
        user.groupProperties?.let { groupProps ->
            try {
                val jsonObject = groupProps.toJSONObject()
                map["groupProperties"] = jsonObject?.toString()
            } catch (e: Exception) {
                Log.e("ExperimentFlutter", "Error encoding groupProperties: ${e.message}")
            }
        }

        return map
    }
}

File: ios/Classes/codec/UserCodec.swift
import AmplitudeExperiment
import Foundation

/**
 * Codec for serializing and deserializing ExperimentUser objects to/from Maps
 * for method channel communication.
 */
enum UserCodec {
    /**
     * Parses an ExperimentUser from a Map representation.
     * @param userMap The map containing user data
     * @return An ExperimentUser object parsed from the map, or nil if userMap is nil
     */
    static func fromMap(_ userMap: [String: Any]?) -> ExperimentUser? {
        guard let userArgs = userMap else {
            return nil
        }

        let builder = ExperimentUserBuilder()

        if let deviceId = userArgs["deviceId"] as? String {
            builder.deviceId(deviceId)
        }
        if let userId = userArgs["userId"] as? String {
            builder.userId(userId)
        }
        if let country = userArgs["country"] as? String {
            builder.country(country)
        }
        if let city = userArgs["city"] as? String {
            builder.city(city)
        }
        if let region = userArgs["region"] as? String {
            builder.region(region)
        }
        if let dma = userArgs["dma"] as? String {
            builder.dma(dma)
        }
        if let language = userArgs["language"] as? String {
            builder.language(language)
        }
        if let platform = userArgs["platform"] as? String {
            builder.platform(platform)
        }
        if let version = userArgs["version"] as? String {
            builder.version(version)
        }
        if let os = userArgs["os"] as? String {
            builder.os(os)
        }
        if let deviceModel = userArgs["deviceModel"] as? String {
            builder.deviceModel(deviceModel)
        }
        if let deviceManufacturer = userArgs["deviceManufacturer"] as? String {
            builder.deviceManufacturer(deviceManufacturer)
        }
        if let carrier = userArgs["carrier"] as? String {
            builder.carrier(carrier)
        }
        if let library = userArgs["library"] as? String {
            builder.library(library)
        }

        // Decode userProperties from JSON string
        if let userPropertiesString = userArgs["userProperties"] as? String,
            let userPropertiesData = userPropertiesString.data(using: .utf8),
            let userProperties = try? JSONSerialization.jsonObject(
                with: userPropertiesData
            ) as? [String: Any]
        {
            builder.userProperties(userProperties)
        }

        // Decode groups from JSON string (Map<String, List<String>>)
        if let groupsString = userArgs["groups"] as? String,
            let groupsData = groupsString.data(using: .utf8),
            let groupsDict = try? JSONSerialization.jsonObject(with: groupsData)
                as? [String: Any]
        {
            var groups: [String: [String]] = [:]
            for (key, value) in groupsDict {
                if let list = value as? [String] {
                    groups[key] = list
                }
            }
            builder.groups(groups)
        }

        // Decode groupProperties from JSON string (Map<String, Map<String, Map<String, Any?>>>)
        if let groupPropertiesString = userArgs["groupProperties"] as? String,
            let groupPropertiesData = groupPropertiesString.data(using: .utf8),
            let groupPropertiesDict = try? JSONSerialization.jsonObject(
                with: groupPropertiesData
            ) as? [String: Any]
        {
            var groupProperties: [String: [String: [String: Any?]]] = [:]
            for (groupKey, groupValue) in groupPropertiesDict {
                if let groupMap = groupValue as? [String: Any] {
                    var innerMap: [String: [String: Any?]] = [:]
                    for (propertyKey, propertyValue) in groupMap {
                        if let propertyMap = propertyValue as? [String: Any] {
                            var finalMap: [String: Any?] = [:]
                            for (key, value) in propertyMap {
                                finalMap[key] = value
                            }
                            innerMap[propertyKey] = finalMap
                        }
                    }
                    groupProperties[groupKey] = innerMap
                }
            }
            builder.groupProperties(groupProperties)
        }

        return builder.build()
    }

    /**
     * Converts a native iOS ExperimentUser object to a Map that can be serialized
     * and sent to Flutter via method channels.
     * @param user The user to convert, or nil
     * @return A map representation of the user, or empty map if user is nil
     */
    static func toMap(_ user: ExperimentUser?) -> [String: Any?] {
        guard let user = user else {
            return [:]
        }

        var map: [String: Any?] = [:]

        if let deviceId = user.deviceId {
            map["deviceId"] = deviceId
        }
        if let userId = user.userId {
            map["userId"] = userId
        }
        if let country = user.country {
            map["country"] = country
        }
        if let city = user.city {
            map["city"] = city
        }
        if let region = user.region {
            map["region"] = region
        }
        if let dma = user.dma {
            map["dma"] = dma
        }
        if let language = user.language {
            map["language"] = language
        }
        if let platform = user.platform {
            map["platform"] = platform
        }
        if let version = user.version {
            map["version"] = version
        }
        if let os = user.os {
            map["os"] = os
        }
        if let deviceModel = user.deviceModel {
            map["deviceModel"] = deviceModel
        }
        if let deviceManufacturer = user.deviceManufacturer {
            map["deviceManufacturer"] = deviceManufacturer
        }
        if let carrier = user.carrier {
            map["carrier"] = carrier
        }
        if let library = user.library {
            map["library"] = library
        }

        // Encode userProperties to JSON string
        if let userProperties = user.userProperties,
            let userPropertiesData = try? JSONSerialization.data(
                withJSONObject: userProperties
            ),
            let userPropertiesString = String(
                data: userPropertiesData,
                encoding: .utf8
            )
        {
            map["userProperties"] = userPropertiesString
        }

        // Encode groups to JSON string (Map<String, List<String>>)
        if let groups = user.groups {
            var groupsAsList: [String: [String]] = [:]
            for (key, value) in groups {
                groupsAsList[key] = Array(value)
            }
            if let groupsData = try? JSONSerialization.data(
                withJSONObject: groupsAsList
            ),
                let groupsString = String(data: groupsData, encoding: .utf8)
            {
                map["groups"] = groupsString
            }
        }

        // Encode groupProperties to JSON string (Map<String, Map<String, Map<String, Any?>>>)
        if let groupProperties = user.groupProperties,
            let groupPropertiesData = try? JSONSerialization.data(
                withJSONObject: groupProperties
            ),
            let groupPropertiesString = String(
                data: groupPropertiesData,
                encoding: .utf8
            )
        {
            map["groupProperties"] = groupPropertiesString
        }

        return map
    }
}
